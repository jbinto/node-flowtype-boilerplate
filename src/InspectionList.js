export default class InspectionList {
  constructor() {
    this.inspections = {}
  }

  add(inspection) {
    console.log('---', inspection, '-----------------------------')
    const {
      // ACTION,
      // AMOUNT_FINED,
      // COURT_OUTCOME,
      ESTABLISHMENT_ADDRESS,
      ESTABLISHMENT_NAME,
      ESTABLISHMENT_STATUS,
      INFRACTION_DETAILS,
      INSPECTION_DATE,
      INSPECTION_ID,
      SEVERITY,
    } = inspection

    // File is in random order; store using a naturally sortable key
    const hashKey = INSPECTION_DATE + INSPECTION_ID

    // Group by inspection ID. Everything is the same except:
    //  (INFRACTION_*, SEVERITY)
    if (this.inspections[hashKey] == null) {
      const summary = {
        name: ESTABLISHMENT_NAME,
        address: ESTABLISHMENT_ADDRESS,
        status: ESTABLISHMENT_STATUS,
        date: INSPECTION_DATE,
        infractions: [],
      }
      this.inspections[hashKey] = summary
    }

    const infraction = {
      details: INFRACTION_DETAILS,
      severity: SEVERITY,
      // action: ACTION,
      // courtOutcome: COURT_OUTCOME,
      // amountFined: AMOUNT_FINED,
    }
    this.inspections[hashKey].infractions.push(infraction)
  }

  count() {
    return Object.keys(this.inspections).length
  }

  toObject() {
    const reverseSortedObject = input => {
      const output = {}
      Object.keys(input)
        .sort()
        .reverse()
        .forEach(key => {
          output[key] = input[key]
        })
      return output
    }

    return reverseSortedObject(this.inspections)
  }

  toArray() {
    return Object.values(this.toObject())
  }
}
