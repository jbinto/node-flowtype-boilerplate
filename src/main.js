import bigXml from 'big-xml'
import _ from 'lodash'
import Case from 'case'

const RECENT_DATE_REGEX = /^2017-0[789]/

const reader = bigXml.createReader('dinesafe.xml', /^ROW$/)

// { tag: 'ROW',
//   text: '\n  \n  \n  \n  \n  \n  \n  \n  \n  \n  \n  \n  \n  \n  \n',
//   children:
//    [ { tag: 'ROW_ID', text: '2489' },
//      { tag: 'ESTABLISHMENT_ID', text: '9003142' },
//      { tag: 'INSPECTION_ID', text: '103851324' },
//      { tag: 'ESTABLISHMENT_NAME', text: 'COACH HOUSE RESTAURANT' },
//      { tag: 'ESTABLISHMENTTYPE', text: 'Restaurant' },
//      { tag: 'ESTABLISHMENT_ADDRESS', text: '574 YONGE ST ' },
//      { tag: 'ESTABLISHMENT_STATUS', text: 'Pass' },
//      { tag: 'MINIMUM_INSPECTIONS_PERYEAR', text: '3' },
//      { tag: 'INFRACTION_DETAILS',
//        text: 'Operator fail to properly wash surfaces in rooms' },
//      { tag: 'INSPECTION_DATE', text: '2016-11-17' },
//      { tag: 'SEVERITY', text: 'M - Minor' },
//      { tag: 'ACTION', text: 'Notice to Comply' },
//      { tag: 'COURT_OUTCOME', text: ' ' },
//      { tag: 'AMOUNT_FINED', text: ' ' } ] }

const inspections = {}

reader.on('record', record => {
  const inspection = _.chain(record.children)
    .keyBy('tag')
    .mapValues('text')
    .value()

  // {
  //   ROW_ID: '27233',
  //   ESTABLISHMENT_ID: '10277628',
  //   INSPECTION_ID: '103956759',
  //   ESTABLISHMENT_NAME: 'LUCKY\'S CORNER',
  //   ESTABLISHMENTTYPE: 'Restaurant',
  //   ESTABLISHMENT_ADDRESS: '2111 JANE ST ',
  //   ESTABLISHMENT_STATUS: 'Conditional Pass',
  //   MINIMUM_INSPECTIONS_PERYEAR: '3',
  //   INFRACTION_DETAILS: 'Operator fail to provide proper equipment',
  //   INSPECTION_DATE: '2017-04-26',
  //   SEVERITY: 'M - Minor',
  //   ACTION: 'Notice to Comply',
  //   COURT_OUTCOME: ' ',
  //   AMOUNT_FINED: ' '
  // }

  const {
    ACTION,
    AMOUNT_FINED,
    COURT_OUTCOME,
    ESTABLISHMENT_ADDRESS,
    ESTABLISHMENT_NAME,
    ESTABLISHMENT_STATUS,
    INFRACTION_DETAILS,
    INSPECTION_DATE,
    INSPECTION_ID,
    SEVERITY,
  } = inspection

  const isFailure = ESTABLISHMENT_STATUS !== 'Pass'
  const isRecent = RECENT_DATE_REGEX.test(INSPECTION_DATE)
  if (!isFailure) return
  if (!isRecent) return

  const HASH_KEY = INSPECTION_DATE + INSPECTION_ID

  // Group by inspection ID. Everything is the same except:
  //  (INFRACTION_*, SEVERITY)
  if (inspections[HASH_KEY] == null) {
    console.log('New inspection ', INSPECTION_ID)
    const summary = {}
    summary.name = ESTABLISHMENT_NAME
    summary.address = ESTABLISHMENT_ADDRESS
    summary.status = ESTABLISHMENT_STATUS
    summary.date = INSPECTION_DATE
    summary.infractions = []
    inspections[HASH_KEY] = summary
  }

  console.log('  infraction ', inspections[HASH_KEY].infractions.length)
  const infraction = {
    details: INFRACTION_DETAILS,
    severity: SEVERITY,
    // action: ACTION,
    // courtOutcome: COURT_OUTCOME,
    // amountFined: AMOUNT_FINED,
  }
  inspections[HASH_KEY].infractions.push(infraction)
})

reader.on('end', () => {
  const reverseSortedObject = input => {
    const output = {}
    Object.keys(input)
      .sort()
      .reverse()
      .forEach(key => {
        output[key] = input[key]
      })
    return output
  }

  const sortedByHashKey = reverseSortedObject(inspections)
  const withoutKeys = Object.values(sortedByHashKey)

  console.log(JSON.stringify(withoutKeys, null, 2))
})
