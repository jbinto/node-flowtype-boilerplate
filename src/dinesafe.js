/* eslint-disable import/prefer-default-export */
import bigXml from 'big-xml'
import _ from 'lodash'
import InspectionList from './InspectionList'

const RECENT_DATE_REGEX = /^2017-0[789]/

export const parseBigXmlToInspection = record =>
  // `big-xml` returns records in the following format:
  //   children:
  //    [ { tag: 'ROW_ID', text: '2489' },
  //      { tag: 'ESTABLISHMENT_ID', text: '9003142' },
  //    ...

  // convert to the following:
  // {
  //   ROW_ID: '27233',
  //   ESTABLISHMENT_ID: '10277628',
  //   ...
  // }
  _.chain(record.children)
    .keyBy('tag')
    .mapValues('text')
    .value()

export const shouldSkip = inspection => {
  const isPass = inspection.ESTABLISHMENT_STATUS === 'Pass'
  const isOld = !RECENT_DATE_REGEX.test(inspection.INSPECTION_DATE)
  return isPass || isOld
}

const processDinesafeXMLtoJSON = (path = './dinesafe.xml') => {
  const inspectionList = new InspectionList()

  const processRecord = record => {
    try {
      const inspection = parseBigXmlToInspection(record)
      if (shouldSkip(inspection)) return
      inspectionList.add(inspection)
      process.stderr.write('.')
    } catch (ex) {
      process.stderr.write('processRecord ERROR', ex)
      process.exit(111)
    }
  }

  const processEnd = () => {
    process.stderr.write(
      `\nOutputting ${inspectionList.count()} failed inspections\n`
    )
    const array = inspectionList.toArray()
    const json = JSON.stringify(array, null, 2)
    process.stdout.write(json.substr(0, 100))
  }

  const reader = bigXml.createReader(path, /^ROW$/)
  reader.on('record', processRecord)
  reader.on('end', processEnd)
}

if (require.main === module) {
  // Running from command line, not reqire/import
  processDinesafeXMLtoJSON()
}
