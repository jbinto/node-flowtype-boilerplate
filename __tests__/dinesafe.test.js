/** @flow */

import { parseBigXmlToInspection, shouldSkip } from '../src/dinesafe'

describe('parseBigXmlToInspection', () => {
  it('has no regressions in snapshot', () => {
    const SAMPLE_OUTPUT_FROM_BIGXML = {
      tag: 'ROW',
      text: '\n  \n  \n  \n  \n  \n  \n  \n  \n  \n  \n  \n  \n  \n  \n',
      children: [
        { tag: 'ROW_ID', text: '781' },
        { tag: 'ESTABLISHMENT_ID', text: '9001121' },
        { tag: 'INSPECTION_ID', text: '103970956' },
        { tag: 'ESTABLISHMENT_NAME', text: "BARBERIAN'S STEAK HOUSE" },
        { tag: 'ESTABLISHMENTTYPE', text: 'Restaurant' },
        { tag: 'ESTABLISHMENT_ADDRESS', text: '7 ELM ST ' },
        { tag: 'ESTABLISHMENT_STATUS', text: 'Pass' },
        { tag: 'MINIMUM_INSPECTIONS_PERYEAR', text: '3' },
        {
          tag: 'INFRACTION_DETAILS',
          text: 'FAIL TO STORE FOOD ON RACKS OR SHELVES O. REG  562/90 SEC. 23',
        },
        { tag: 'INSPECTION_DATE', text: '2017-05-17' },
        { tag: 'SEVERITY', text: 'S - Significant' },
        { tag: 'ACTION', text: 'Corrected During Inspection' },
        { tag: 'COURT_OUTCOME', text: ' ' },
        { tag: 'AMOUNT_FINED', text: ' ' },
      ],
    }

    const result = parseBigXmlToInspection(SAMPLE_OUTPUT_FROM_BIGXML)
    expect(result).toMatchSnapshot()
  })
})

describe('shouldSkip', () => {
  it('should skip an older date', () => {
    expect(
      shouldSkip({
        INSPECTION_DATE: '2016-01-01',
      })
    ).toEqual(true)
  })

  it('should skip a Pass', () => {
    expect(
      shouldSkip({
        ESTABLISHMENT_STATUS: 'Pass',
      })
    ).toEqual(true)
  })

  it('should not skip a recent failure', () => {
    expect(
      shouldSkip({
        ESTABLISHMENT_STATUS: 'Conditional Pass',
        INSPECTION_DATE: '2017-08-01',
      })
    ).toEqual(false)
  })
})
