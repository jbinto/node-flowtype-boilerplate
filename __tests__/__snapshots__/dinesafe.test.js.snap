// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseBigXmlToInspection has no regressions in snapshot 1`] = `
Object {
  "ACTION": "Corrected During Inspection",
  "AMOUNT_FINED": " ",
  "COURT_OUTCOME": " ",
  "ESTABLISHMENTTYPE": "Restaurant",
  "ESTABLISHMENT_ADDRESS": "7 ELM ST ",
  "ESTABLISHMENT_ID": "9001121",
  "ESTABLISHMENT_NAME": "BARBERIAN'S STEAK HOUSE",
  "ESTABLISHMENT_STATUS": "Pass",
  "INFRACTION_DETAILS": "FAIL TO STORE FOOD ON RACKS OR SHELVES O. REG  562/90 SEC. 23",
  "INSPECTION_DATE": "2017-05-17",
  "INSPECTION_ID": "103970956",
  "MINIMUM_INSPECTIONS_PERYEAR": "3",
  "ROW_ID": "781",
  "SEVERITY": "S - Significant",
}
`;
